### PR TITLE
feat: slack client to support multiple org

### PR DESCRIPTION
**Pull request by ekremney**
Please ensure your pull request adheres to the following guidelines:
- [ ] make sure to link the related issues in this description
- [ ] when merging / squashing, make sure the fixed issue references are visible in the commits, for easy compilation of release notes

## Related Issues

Thanks for contributing!

### CODE REVIEW

**Review by dzehnder — APPROVED** [review]

File: packages/spacecat-shared-slack-client/src/index.js [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+      .reduce((acc, pair) => {
+        acc[pair.target] = new WebClient(pair.token);
+        return acc;
+      }, {});
+    this.log = log;
+  }
+
+  #getClient(target) {
+    if (!this.clients[target]) {
+      throw new Error(`Environment variable '${ENV_PREFIX}${target}' does not exist. Slack Client could not initialized.`);
```
Comment (solaris007) [inline packages/spacecat-shared-slack-client/src/index.js:63]: `could not *be* ...` or `failed to initialize`

File: packages/spacecat-shared-slack-client/src/index.js [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+const DEFAULT_PARAMS = {
+  unfurl_links: false,
+  unfurl_media: false,
+};
+
+export class SlackClient {
+  static createFrom(context) {
+    if (context.slackClient) return context.slackClient;
+
+    const expected = Object.values(SLACK_TARGETS).map((target) => `${ENV_PREFIX}${target}`);
```
Comment (solaris007) [inline packages/spacecat-shared-slack-client/src/index.js:30]: i suggest wrapping `${ENV_PREFIX}${target}` into a little function, it is used multiple times

File: packages/spacecat-shared-slack-client/src/index.js [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+
+    if (targetTokenPairs.length === 0) {
+      throw Error('No environment variable containing a slack token found');
+    }
+
+    this.clients = targetTokenPairs
+      .reduce((acc, pair) => {
+        acc[pair.target] = new WebClient(pair.token);
+        return acc;
+      }, {});
```
Comment (solaris007) [inline packages/spacecat-shared-slack-client/src/index.js:57]: could use destructuring:
```suggestion
      .reduce((acc, { target, token }) => {
        acc[target] = new WebClient(token);
        return acc;
      }, {});
```

File: packages/spacecat-shared-slack-client/src/index.js [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+        ...DEFAULT_PARAMS,
+        ...opts,
+      });
+
+      return {
+        channelId: result.channel,
+        threadId: result.ts,
+      };
+    } catch (error) {
+      this.log.error(`Slack message failed to send to ${target}`);
```
Comment (solaris007) [inline packages/spacecat-shared-slack-client/src/index.js:81]: please include more details, i.e. message and/or stack trace

File: packages/spacecat-shared-slack-client/src/index.d.ts [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+   */
+  constructor(targetTokenPairs: TargetTokenPair[]);
+
+  /**
+   * Asynchronous method to create a RUM backlink.
+   * @param {string} target - A string representing the target slack workspace/org
+   * @param {number} opts - An object containing the data sent to Slack API
+   * @returns A Promise resolving to an object containing the data returned by Slack API
+   */
+  postMessage(target: string, opts: object): Promise<object>;
```
Comment (solaris007) [inline packages/spacecat-shared-slack-client/src/index.d.ts:45]: i would simplify the signature to postMessage(message, opts) - most people want to send a simple message and don't want to deal with knowing slack internal options, which still can be optionally provided...

**Review by solaris007 — COMMENTED** [review]
i'd propose to split the client-factory part from the actual slack client.  this aligns with the principle of single responsibility, ensuring that each client instance is dedicated to a specific target (i.e., a Slack workspace). This redesign would simplify the client interface by removing the need to specify the target for every public method call.

Incomplete example:
```js
class SlackClientFactory {
  constructor(log) {
    this.clients = {};
    this.log = log;
  }

  // createFrom / lambda context handling here...

  getClientForTarget(target) {
    if (!this.clients[target]) {
      const token = this.getTokenForTarget(target);
      if (!token) {
        throw new Error(`No token found for target: ${target}`);
      }
      this.clients[target] = new SlackClient(token, this.log);
    }
    return this.clients[target];
  }

  getTokenForTarget(target) {
    return process.env[`${ENV_PREFIX}${target}`];
  }
}

class SlackClient {
  constructor(token, log) {
    this.client = new WebClient(token);
    this.log = log;
  }

  async postMessage(message, opts) {
    try {
      const result = await this.client.chat.postMessage({
        ...DEFAULT_PARAMS,
        text: message,
        ...opts,
      });

      return {
        channelId: result.channel,
        threadId: result.ts,
      };
    } catch (error) {
      this.log.error('Slack message failed to send');
      throw error;
    }
  }
}
```

File: packages/spacecat-shared-slack-client/src/index.d.ts (hunk 5 above)
Comment (dbrrt) [inline packages/spacecat-shared-slack-client/src/index.d.ts:45]:
```suggestion
  postMessage(target: string, opts?: object): Promise<object>;
```

**Comment by ekremney** [comment]
I incorporated all the reviews and added the file upload method. Please re-review

File: packages/spacecat-shared-slack-client/package.json [hunk 6]
```diff
@@ -0,0 +1,48 @@
+{
+  "name": "@adobe/spacecat-shared-slack-client",
+  "version": "1.2.1",
```
Comment (solaris007) [inline packages/spacecat-shared-slack-client/package.json:3]: are we already at this version? should probably be reset.

File: packages/spacecat-shared-slack-client/package.json [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+  "publishConfig": {
+    "access": "public"
+  },
+  "dependencies": {
+    "@adobe/fetch": "4.1.1",
+    "@adobe/helix-shared-wrap": "2.0.0",
+    "@adobe/helix-universal": "4.4.1",
+    "@adobe/spacecat-shared-utils": "1.7.2",
+    "@slack/web-api": "7.0.1",
+    "aws4": "1.12.0"
```
Comment (solaris007) [inline packages/spacecat-shared-slack-client/package.json:37]: i think that should be removed

File: packages/spacecat-shared-slack-client/package.json [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+    "url": "https://github.com/adobe/spacecat-shared/issues"
+  },
+  "homepage": "https://github.com/adobe/spacecat-shared#readme",
+  "publishConfig": {
+    "access": "public"
+  },
+  "dependencies": {
+    "@adobe/fetch": "4.1.1",
+    "@adobe/helix-shared-wrap": "2.0.0",
+    "@adobe/helix-universal": "4.4.1",
```
Comment (solaris007) [inline packages/spacecat-shared-slack-client/package.json:34]: those should probably be removed.

helix-universal - is it worth it just for the type def?

File: packages/spacecat-shared-slack-client/src/index.js [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+    return {
+      channelId: result.channel,
+      threadId: result.ts,
+    };
+  }
+
+  async fileUpload(file) {
+    const result = await this.#apiCall('files.uploadV2', file);
+
+    if (result.files.length === 0) {
```
Comment (solaris007) [inline packages/spacecat-shared-slack-client/src/index.js:76]: are you sure `result.files` is always defined? I've seen alternate responses in the doc like:
```
{
    "ok": false,
    "error": "invalid_auth"
}
```

**Review by solaris007 — APPROVED** [review]
LGTM, only minor comments

File: packages/spacecat-shared-slack-client/src/index.js (hunk 9 above)
Comment (ekremney) [reply packages/spacecat-shared-slack-client/src/index.js:76]: when `ok` is not `true`, the `WebClient` throws an error.

File: packages/spacecat-shared-slack-client/package.json (hunk 8 above)
Comment (ekremney) [reply packages/spacecat-shared-slack-client/package.json:34]: I'd rather keep it for the type defs

**Review by solaris007 — APPROVED** [review]

**Pull request merged**